### ISSUE TITLE
Deleting replication controllers on HEAD & 1.1 masters leaves pods running

### ISSUE THREAD

**Issue by ihmccreery** [kubernetes/kubernetes#15123]
It seems that HEAD masters can't delete pods on 1.0 nodes.

After spinning up a (GKE) cluster at 1.0.6 and upgrading the master to `1.2.0-alpha.1.455+4b7329ad8c52f6-dirty`, I can't delete pods:

```
$ kubectl get pods
NAME                     READY     STATUS    RESTARTS   AGE
$ kubectl create -f ./docs/user-guide/walkthrough/replication-controller.yaml 
replicationcontrollers/nginx-controller
$ kubectl get pods
NAME                     READY     STATUS    RESTARTS   AGE
nginx-controller-l4rhx   0/1       Running   0          8s
nginx-controller-tfwpp   0/1       Running   0          8s
$ kubectl delete -f ./docs/user-guide/walkthrough/replication-controller.yaml                                                                                        
replicationcontrollers/nginx-controller
$ kubectl get pods
NAME                     READY     STATUS    RESTARTS   AGE
nginx-controller-l4rhx   1/1       Running   0          6m
nginx-controller-tfwpp   1/1       Running   0          6m
```

I uncovered this because there are several e2e tests failing on version-skewed clusters, and they seem to be all having problems deleting pods.

This is likely related to:

- Kubectl e2e tests leave resources running after stop during upgrade #15112
- DNS e2es are failing on version-skewed clusters, where master is at 1.1 and nodes are at 1.0 #15120

**Comment by ihmccreery** [comment]
This doesn't actually seem to be effecting 1.1.  Moving down to P1.

**Comment by ihmccreery** [comment]
Lies.  Bumping back up to P0, because this does seem to be affecting 1.1.

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"0", GitVersion:"v1.0.6", GitCommit:"388061f00f0d9e4d641f9ed4971c775e1654579d", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"1+", GitVersion:"v1.1.0-beta.237+be47a12477d436-dirty", GitCommit:"be47a12477d436b27b3ed1bd1056a9ac6a641658", GitTreeState:"dirty"}
$ kubectl describe node gke-pods-check-ed00f170-node-76v2
Name:                   gke-pods-check-ed00f170-node-76v2
...
Version:
 Kernel Version:                3.16.0-0.bpo.4-amd64
 OS Image:                      Debian GNU/Linux 7 (wheezy)
 Container Runtime Version:     docker://1.6.2
 Kubelet Version:               v1.0.6
 Kube-Proxy Version:            v1.0.6
$ kubectl create -f ./redis-master-controller.yaml 
replicationcontrollers/redis-master
$ kubectl get pods
NAME                 READY     STATUS    RESTARTS   AGE
redis-master-zrtnz   0/1       Pending   0          6s
$ kubectl stop -f ./redis-master-controller.yaml                                                                           
replicationcontrollers/redis-master
$ kubectl get pods
NAME                 READY     STATUS    RESTARTS   AGE
redis-master-zrtnz   1/1       Running   0          29s
$ kubectl get pods
NAME                 READY     STATUS    RESTARTS   AGE
redis-master-zrtnz   1/1       Running   0          4m
```

Also see the *kubernetes-upgrade-1.0-1.1-gke-step3-e2e-old* job.  There are a lot of failing tests, most of which seem to be failing because pods aren't deleting.

**Comment by ihmccreery** [comment]
Assigning to @thockin; please reassign as necessary.

**Comment by ihmccreery** [comment]
A lot of the tests in Jenkins are also failing after the nodes have been upgraded.  My guess is that the pods are still stuck from the first e2e run during the second, so I think this is just a version-skew issue, but I could be wrong.

**Comment by mikedanese** [comment]
cc @lavalamp

**Comment by bgrant0607** [comment]
cc @smarterclayton @derekwaynecarr @ncdc Could we please get some help root-causing this? One problem has been identified (#15895), but it's believed to not be the only problem here.

@ihmccreery What version of kubectl did you use?

**Comment by smarterclayton** [comment]
You can delete pods with --grace-period=0

**Comment by mikedanese** [comment]
The reason why we don't think this is fully explained by #15895 is because the last list pods shows running pods as opposed to terminating pods.

**Comment by ihmccreery** [comment]
Also, just telling all of our users, "use `--grace-period=0` or else your deleted pods will live forever" is not a good UX.  This needs to work out-of-the-box.

> @ihmccreery What version of kubectl did you use?

Both Jenkins and I are running `kubectl` v1.0.6:

```
03:41:57 Client Version: version.Info{Major:"1", Minor:"0", GitVersion:"v1.0.6", GitCommit:"388061f00f0d9e4d641f9ed4971c775e1654579d", GitTreeState:"clean"}
03:41:57 Server Version: version.Info{Major:"1", Minor:"1+", GitVersion:"v1.1.0-beta.249+2ffc8202883231-dirty", GitCommit:"2ffc82028832314dd1202db54e5b4b1b3c8ac5d9", GitTreeState:"dirty"}
```

**Comment by ncdc** [comment]
I just tested on my local vagrant VM. I built the latest master and ran `hack/local-up-cluster.sh`. I manually killed the kubelet process. I downloaded the 1.0.6 release tarball and started the kubelet by hand just like `hack/local-up-cluster.sh` would. I created the rc as described above and then deleted it. My pods show Terminating (i.e. what you see in #15895), not Running. Perhaps there is a race somewhere or I somehow didn't reproduce this in the exact same manner.

**Comment by smarterclayton** [comment]
No one was proposing that.

On Oct 19, 2015, at 8:00 PM, Isaac Hollander McCreery <
notifications@github.com> wrote:

Also, just telling all of our users, "use --grace-period=0 or else your
deleted pods will live forever" is not a good UX. This needs to work
out-of-the-box.

@ihmccreery <https://github.com/ihmccreery> What version of kubectl did you
use?

Both Jenkins and I are running kubectl v1.0.6:

03:41:57 Client Version: version.Info{Major:"1", Minor:"0",
GitVersion:"v1.0.6",
GitCommit:"388061f00f0d9e4d641f9ed4971c775e1654579d",
GitTreeState:"clean"}
03:41:57 Server Version: version.Info{Major:"1", Minor:"1+",
GitVersion:"v1.1.0-beta.249+2ffc8202883231-dirty",
GitCommit:"2ffc82028832314dd1202db54e5b4b1b3c8ac5d9",
GitTreeState:"dirty"}

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/15123#issuecomment-149380291>
.

**Comment by smarterclayton** [comment]
Was the client 1.0 or 1.1?

On Oct 19, 2015, at 7:56 PM, Mike Danese <notifications@github.com> wrote:

The reason why we don't think this is fully explained by #15895
<https://github.com/kubernetes/kubernetes/issues/15895> is because the last
list pods shows running pods as opposed to terminating pods.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/15123#issuecomment-149379673>
.

**Comment by mikedanese** [comment]
@smarterclayton 1.0 client.

**Comment by ncdc** [comment]
My test was with a 1.1 client. I'll retest with 1.0.

**Comment by smarterclayton** [comment]
Terminating is not a new phase (not backwards compatible), so old clients
don't know that deletion has been requested but not yet completed.

On Oct 19, 2015, at 9:12 PM, Mike Danese <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> 1.0 client.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/15123#issuecomment-149390846>
.

**Comment by ncdc** [comment]
Ok, I see the same behavior. On a 1.0 client, the status is running. On a HEAD client, it's terminating. So is this essentially a dupe of 15895?

**Comment by smarterclayton** [comment]
It's more of a "old clients don't see new behavior". It's not really a bug.

**Comment by ihmccreery** [comment]
> No one was proposing that.

Acknowledged.  I misunderstood your comment.

> Terminating is not a new phase (not backwards compatible), so old clients don't know that deletion has been requested but not yet completed.

Can we still call this behavior `v1` then?  @lavalamp, you've mentioned concern over version-skew problems between clients and the API.  Any strong feelings?

**Comment by davidopp** [comment]
> It's more of a "old clients don't see new behavior". It's not really a bug.

Just to make sure I understand: you do agree that the fact that the pods never go away is a bug, but the fact that they show Running instead of Terminating is not a bug?

(And perhaps you are further implying that because Running vs. Terminating is a red herring, this probably actually is a dupe of #15895?)

**Comment by smarterclayton** [comment]
On Oct 20, 2015, at 12:30 AM, David Oppenheimer <notifications@github.com>
wrote:

It's more of a "old clients don't see new behavior". It's not really a bug.

Just to make sure I understand: you do agree that the fact that the pods
never go away is a bug, but the fact that they show Running instead of
Terminating is not a bug?


Yes, node controller should identify that a Kubelet cannot perform its 1.1
duties and manage that for it.  1.0 Kubectl does not see that a pod is in
the process of deleting because we cannot add new phases without breaking
backwards compatibility and a new condition would have to be set by the
Kubelet (since it blanket overwrites the pod status, and a 1.0 Kubelet
would not set the condition anyway).

If the node controller was picking up the slack for the Kubelet, you'd see
(as a client) a delete request that is acknowledged (200) and the pod goes
away a short time later.  If the cluster was fully 1.1, you'd see your
delete accepted and then go away after the grace period.





(And perhaps you are further implying that because Running vs. Terminating
is a red herring, this probably actually is a dupe of #15895
<https://github.com/kubernetes/kubernetes/issues/15895>?)

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/15123#issuecomment-149428869>
.

**Comment by davidopp** [comment]
I'm assigning this to @mikedanese since the consensus now seems to be that this is the same as #15895

**Comment by davidopp** [comment]
We closed #15895 so I'm going to close this one. If this turns out to have been caused by something else ,we can reopen.
